<a href="https://colab.research.google.com/github/Lalo287/Simulacion_II-Tareas/blob/main/Parcial%202/Tarea_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Líneas de espera

##Tarea 4.


Convierta los códigos vistos en clase (para calcular W, L y P(N=k)) en funciones y úselos para responder todo lo que sea posible mediante simulaciones. Luego verifique sus resultados utilizando las fórmulas de la teoría.

Si no es posible usar la simulación para responder alguna pregunta en particular, argumente porqué y respóndala teóricamente.

### **En la biblioteca**

Un encargado de la biblioteca es el único en el mostrador durante todo su turno de trabajo.
Las llegadas al mostrador siguen una distribución de Poisson con una media de 8 por hora.
Cada usuario de la biblioteca es atendido de uno en uno, y el tiempo de servicio sigue una
distribución exponencial con una media de 5 minutos.

1.   ¿Cuál es la probabilidad de que se forme cola?
2.   ¿Cuál es la longitud media de la cola?
3.   ¿Cuál es el tiempo medio que un cliente pasa en la biblioteca hasta que le han atendido?
4.   ¿Cuál es el tiempo medio que un cliente pasa en la cola esperando a que le atiendan?
5.   El encargado de la biblioteca pasa su tiempo en que no hay clientes clasificando artículos
de revistas. Si puede clasificar 22 fichas por hora como media cuando trabaja continuamente,
¿Cuántas fichas puede ordenar durante su trabajo?



In [7]:
from random import expovariate
from statistics import mean
from math import inf as Infinity
from random import seed
from matplotlib.pyplot import *

In [54]:
# Parametros

lamda = 8                         # Razón de arrivos (1/Lambda)  Ojo
mu = 12                            # Razón de salida (1/Mu)       Ojo


Num_Pkts = 100000                   # Numero de paquetes a simular
#T = 100000
count = 0                           # Contador de paquetes simulados
t = 0
N = 0                               # Variable de estado: numero de paquetes en el sistema

tll = expovariate(lamda)            # Tiempo de llegada
ts = Infinity                       # Tiempo de salida
t_evento_anterior = 0.0             # tiempo del último evento ****************Nuevas líneas de código***********

# Variables de salida
tll_Data = []                       # Lista de tiempos de llegada
ts_Data = []                        # Lista de tiempos de salida
r_Data = []                         # Lista de retrasos individuales

Data = {}                           # Diccionario *****************************************************************

while count < Num_Pkts:
#while t < T:
    if tll < ts:                           # Evento de llegada
        t = tll
        
        # longitud del intervalo entre llegadas *********************************************************************
        delta = t - t_evento_anterior
        if N in Data: 
            Data[N] += delta
        else:
            Data[N] = delta
        t_evento_anterior = t
        
        tll_Data.append(t)
        N = N + 1.0
        tll = t + expovariate(lamda) 
        if N == 1:
            ts = t + expovariate(mu)
    else:                                         # Evento de salida
        t = ts
        
        # Longitud del intervalo de tiempo **************************************************************************
        delta = t - t_evento_anterior
        if N in Data:
            Data[N] += delta
        else:
            Data[N] = delta
        t_evento_anterior = t
        
        ts_Data.append(t)
        N = N - 1.0
        count = count + 1                         # Paquetes simulados
        if N > 0:
            ts = t + expovariate(mu)
        else:
            ts = Infinity

##################################### se procesan los datos de salida ########################################
            
# Estimación del retraso promedio:

for i in range(Num_Pkts):
#for i in range(T):
    d = ts_Data[i] - tll_Data[i]
    r_Data.append(d)

# Estimación de la probabilidad de P(N=k):
for (key, value) in Data.items():
    Data[key] = value/t

media = 0.0

1. ¿Cuál es la probabilidad de que se forme cola? 
$$    $$
$$  P_{1<12} = 1-P_0+P_1 $$
$$    $$

In [55]:
P01 = 0

for (key, value) in Data.items():
  
  if key <= 1:
    P01 = P01 + round(value, 4) 

print("P_1<12 = ", round(1-P01,4))
    

P_1<12 =  0.4407


Entonces:
$$  P_{1<12} = 0.4453 = 44.53 $$
$$    $$

2.   ¿Cuál es la longitud media de la cola?
$$      $$
$$ρ=\frac{λ}{μ}=\frac{8}{12}=\frac{2}{3}$$

Donde $L_q$

$$  L_q = \frac{ρ^2}{1-ρ} $$
$$    $$ 

In [52]:
p = lamda/mu

Lq = (p*p)/(1-p)
print( "Cantidad promedio de clientes que se encuentran en la fila =   Lq =", round( Lq, 4) )

Cantidad promedio de clientes que se encuentran en la fila =   Lq = 1.3333


Entonces:
$$    $$
$$  L_q = 1.333 $$
$$    $$


3.   ¿Cuál es el tiempo medio que un cliente pasa en la biblioteca hasta que le han atendido?
$$    $$
$$ W = \frac{ρ}{\lambda(1-ρ)}$$
$$    $$

In [58]:
W = (p)/(lamda*(1-p))

print( "Tiempo de espera en el sistema =    W = ", round( W, 4) )

Tiempo de espera en el sistema =    W =  0.25


Entonces:
$$    $$
$$  W = 0.25 $$
$$    $$


4.   ¿Cuál es el tiempo medio que un cliente pasa en la cola esperando a que le atiendan?

$$    $$
$$ W_q = \frac{ρ^2}{\lambda(1-ρ)}$$
$$    $$

In [56]:
Wq = (p*p)/(lamda*(1-p))

print( "Tiempo de espera en la cola =    Wq = ", round( Wq, 4) )

Tiempo de espera en la cola =    Wq =  0.1667


Entonces:
$$    $$
$$  W_q = 0.1667 $$
$$    $$


### **Lavado de autos**
Un pequeño autoservicio de lavado que solo tienen capacidad para atender de uno en uno a los autos que llegan, tiene un aparcamiento para 10 coches, incluyendo el que está siendo lavado. La empresa ha estimado que los coches llegan siguiendo una distribución de Poisson con una media de 20 coches/hora, el tiempo de servicio sigue una distribución exponencial de 12 minutos. La empresa abre durante 10 horas al día.


¿Cuál es la media de coches perdidos cada día debido a las limitaciones de espacio?

In [61]:
from math import log
from random import random, expovariate
from statistics import mean

def p(x):
  return -(1/x)*log(random())
 
def lc():
  T = 10                  #Tiempo de la simulación
  l = 20                  #Parámetro del proceso de entrada
  m = 5                   #Parámetro del proceso de salida
  inf = 1000000     
  t = 0                   #Tiempo actual del sistema
  te = 0                  #Tiemmpo de entrada
  ts = 0                  #Timepo de salida
  ta = 0          
  k = 0                   #Conatdor de coches que se fueron
  c = 0                   #Contador de clientes en el sistema

  while t<T or c>0:
    if t<T:               #Si el tiempo actual supera al tiempo total de atención, el timepo de entrada será infinito 
      te = p(l)           #De esta forma evitamos entradas después del horario establecido
    else:
      te = inf

    ts = p(m)
    if c == 0:             #Si no hay clientes en el sistema, el timepo de salida es infinito para evitar salidas ilógicas
      ts = inf

    ta = min(te, ts) 

    if(ta == te):
      t = t + te
      if c <10:
        c = c + 1
        #print("%.4f"%t,c)
      else:
        k = k + 1
        #print("Coche sin entrar")

    else:
      t = t + ts
      c = c - 1
      #print("           %.4f"%t,c)
  return k

it = 200000
K = []

for i in range(it):
  K.append(lc())

r = mean(K)

print(" Coches en promedio que no fueron atendidos: ", r )

 Coches en promedio que no fueron atendidos:  141.25513




> **Entonces en promedio no fueron atendidos = 141.25 coches**



### **John Macko**

John Macko es alumno en la U de Ozark. Hace trabajos extraños para aumentar sus ingresos.
Las peticiones de trabajo llegan en promedio cada 5 días, pero el tiempo entre ellas es exponencial. El tiempo para terminar un trabajo también es exponencial, con una media de 4 días.

1.  ¿Cuál es la probabilidad de que le falte trabajo a John?
2.  Si John cobra unos 50 por cada trabajo, ¿cuál es su ingreso mensual promedio?
3.  Si al final del semestre John decide subcontratar los trabajos pendientes a $40 cada uno, ¿cuánto debe esperar pagar en promedio?

In [66]:
# Parametros

lamda = 1/5                         # Razón de arrivos (1/Lambda)  Ojo
mu = 1/4                            # Razón de salida (1/Mu)       Ojo


Num_Pkts = 100000                   # Numero de paquetes a simular
#T = 100000
count = 0                           # Contador de paquetes simulados
t = 0
N = 0                               # Variable de estado: numero de paquetes en el sistema

tll = expovariate(lamda)            # Tiempo de llegada
ts = Infinity                       # Tiempo de salida
t_evento_anterior = 0.0             # tiempo del último evento ****************Nuevas líneas de código***********

# Variables de salida
tll_Data = []                       # Lista de tiempos de llegada
ts_Data = []                        # Lista de tiempos de salida
r_Data = []                         # Lista de retrasos individuales

Data = {}                           # Diccionario *****************************************************************

while count < Num_Pkts:
#while t < T:
    if tll < ts:                           # Evento de llegada
        t = tll
        
        # longitud del intervalo entre llegadas *********************************************************************
        delta = t - t_evento_anterior
        if N in Data: 
            Data[N] += delta
        else:
            Data[N] = delta
        t_evento_anterior = t
        
        tll_Data.append(t)
        N = N + 1.0
        tll = t + expovariate(lamda) 
        if N == 1:
            ts = t + expovariate(mu)
    else:                                         # Evento de salida
        t = ts
        
        # Longitud del intervalo de tiempo **************************************************************************
        delta = t - t_evento_anterior
        if N in Data:
            Data[N] += delta
        else:
            Data[N] = delta
        t_evento_anterior = t
        
        ts_Data.append(t)
        N = N - 1.0
        count = count + 1                         # Paquetes simulados
        if N > 0:
            ts = t + expovariate(mu)
        else:
            ts = Infinity

##################################### se procesan los datos de salida ########################################
            
# Estimación del retraso promedio:

for i in range(Num_Pkts):
#for i in range(T):
    d = ts_Data[i] - tll_Data[i]
    r_Data.append(d)



# Estimación de la probabilidad de P(N=k):
for (key, value) in Data.items():
    Data[key] = value/t

# Se verifica que la probabilidad total es 1.0

# Se verifican los valores de expectación
media = 0.0

for (key, value) in Data.items():
    media = media + key*value
    

1. ¿Cuál es la probabilidad de que le falte trabajo a John?
$$    $$
$$  P_0 = 1-ρ $$
$$    $$

In [71]:
P0 = 0

for (key, value) in Data.items():
  
  if key == 0:
    P0 = P0 + round(value, 4) 

print("P_0 = ", round(P0,4))
    

P_0 =  0.205


Entonces:
$$  P_0 = 0.205 = 20.5 $$
$$    $$